k8s/core/configmap/configmap.go

```diff
@@ -1,188 +1,101 @@
 package configmap
 
 import (
-	"fmt"
-	"strings"
 	"time"
 
-	"github.com/portworx/sched-ops/k8s/core"
-	"github.com/sirupsen/logrus"
-	corev1 "k8s.io/api/core/v1"
-	k8s_errors "k8s.io/apimachinery/pkg/api/errors"
-	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"github.com/google/uuid"
+	log "github.com/sirupsen/logrus"
 )
 
-// New returns the ConfigMap interface. It also creates a new
-// configmap in k8s for the given name if not present and puts the data in it.
-func New(
-	name string,
-	data map[string]string,
-	lockTimeout time.Duration,
-	lockAttempts uint,
-	v2LockRefreshDuration time.Duration,
-	v2LockK8sLockTTL time.Duration,
-) (ConfigMap, error) {
-	if data == nil {
-		data = make(map[string]string)
-	}
+func (c *configMap) Instance() *coreConfigMap {
 
-	labels := map[string]string{
-		configMapUserLabelKey: TruncateLabel(name),
-	}
-	data[pxOwnerKey] = ""
-
-	cm := &corev1.ConfigMap{
-		ObjectMeta: meta_v1.ObjectMeta{
-			Name:      name,
-			Namespace: k8sSystemNamespace,
-			Labels:    labels,
-		},
-		Data: data,
+	if c.pxNs == c.config.nameSpace {
+		//fresh install ot upgrade completed
+		return c.config
 	}
 
-	if _, err := core.Instance().CreateConfigMap(cm); err != nil &&
-		!k8s_errors.IsAlreadyExists(err) {
-		return nil, fmt.Errorf("failed to create configmap %v: %v",
-			name, err)
-	}
+	existingConfig := c.config
+	c.copylock.Lock(uuid.New().String())
+	defer c.copylock.Unlock()
 
-	if v2LockK8sLockTTL == 0 {
-		v2LockK8sLockTTL = v2DefaultK8sLockTTL
+	lockMap, err := c.copylock.get()
+	if err != nil {
+		log.Errorf("Error during fetching data from copy lock %s", err)
+		return existingConfig
 	}
+	status := lockMap[upgradeCompletedStatus]
+	if status == trueString {
```

Comment:
How are we handling IN_PROGRESS state.
Let's say N1 is doing the copy. At the same time N2 is trying to update some value usingUpdate() function. According to this implementation, on N2 existingConfig will be returned and could be updating while N1 could be deleting the whole configmap at line 52. Don't we need to wait if a copy is IN_PROGRESS?